include/dali/core/bitmask.h

```diff
@@ -0,0 +1,231 @@
+    }
+  }
+
+  /**
+   * @brief Fill the entire mask with given value.
+   */
+  void fill(bool value) {
+    size_t s = size();
+    clear();
+    resize(s, value);
+  }
+
+  /**
+   * @brief A pseudoreference to a single bit in the mask
+   */
+  struct bitref {
+    bitref &operator=(bool value) {
+      if (value)
+        *storage |= mask;
+      else
+        *storage &= ~mask;
+      return *this;
+    }
+    constexpr operator bool() const {
+      return *storage & mask;
+    }
+   private:
+    friend class bitmask;
+    constexpr bitref(bit_storage_t *array, ptrdiff_t index)
+    : storage(&array[index >> storage_bits_log])
+    , mask(bit_storage_t(1) << (index & (storage_bits - 1))) {}
+
+    bit_storage_t *storage;
+    bit_storage_t mask;
+  };
+
+  using reference_type = bitref;
+  using value_type = bool;
+
+  void push_back(bool value) {
+    size_t req_words = (size_ + storage_bits) >> storage_bits_log;
+    if (req_words > storage_.size())
+      storage_.push_back(value ? 1 : 0);  // put LSB in the storage vector
+    else
+      (*this)[size_] = value;  // set the last bit
+    size_++;
+  }
+
+  void pop_back() {
+    assert(!empty());
+    (*this)[--size_] = 0;
+    if ((size_ & (storage_bits - 1)) == 0)
+      storage_.pop_back();
+  }
+
+  constexpr bool empty() const noexcept { return size_ == 0; }
+
+  void reserve(size_t capacity) {
+    storage_.reserve((capacity + storage_bits - 1) >> storage_bits_log);
+  }
+
+  size_t capacity() const noexcept {
+    return storage_.capacity() * storage_bits;
+  }
+
+  void resize(size_t new_size, bool value = false) {
+    if (new_size == size())
+      return;
+    bit_storage_t ones = ~bit_storage_t(0);
+    bit_storage_t fill = value ? ones : 0;
+    size_t prev_words = storage_.size();
+    storage_.resize((new_size + storage_bits - 1) >> storage_bits_log, fill);
+    if (storage_.size() == prev_words && new_size > size()) {
+      if (value) {
+        storage_.back() |= ones << (size_ & (storage_bits - 1));
+      }
+    }
+    if (new_size & (storage_bits - 1)) {
+      bit_storage_t mask = ones;
+      mask = (mask << (new_size & (storage_bits - 1)));  // remove leading ones
+      mask = ~mask;  // flip values to keep only the leading ones
+      storage_.back() &= mask;
+    }
+    size_ = new_size;
+  }
+
+  void clear() {
+    storage_.clear();
+    size_ = 0;
+  }
+
+  bit_storage_t *data() noexcept {
+    return storage_.data();
+  }
+
+  const bit_storage_t *data() const noexcept {
+    return storage_.data();
+  }
+
+  constexpr size_t size() const noexcept {
+    return size_;
+  }
+  constexpr ssize_t ssize() const noexcept {
+    return size_;
+  }
+
+  bitref operator[](ptrdiff_t index) {
+    return { storage_.data(), index };
+  }
+
+  bool operator[](const ptrdiff_t index) const {
+    bit_storage_t storage = storage_[index >> storage_bits_log];
+    bit_storage_t mask = bit_storage_t(1) << (index & (storage_bits - 1));
+    return storage & mask;
```

Comment:
Can't you express this using bitref as well?

Reply:
No, because bitref requires non-const pointer.